File: java/src/main/java/vsdk/toolkit/common/dataStructures/CircularList.java
package vsdk.toolkit.common.dataStructures;

/**
Implementation is based on Stroustrup 'The C++ Programming Language' Section 8.3
*/

public class CircularList<T> extends CircularListBase {
    public void add(T data) {
        addLink(new CircularListNode<>(data));
    }

    public void append(T data) {
        appendLink(new CircularListNode<>(data));
    }

    public void removeAll() {
        CircularListNode<T> link = (CircularListNode<T>)remove();
        while (link != null) {
            link.nextLink = null;
            link = (CircularListNode<T>)remove();
        }
    }

    @Override
    public void clear() {
        super.clear();
    }

    public CircularListBase baseList() {
        return this;
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/dataStructures/CircularList.ts
import { CircularListBase } from "./CircularListBase";
import { CircularListNode } from "./CircularListNode";

export class CircularList<T> extends CircularListBase {
  public add(data: T): void {
    this.addLink(new CircularListNode<T>(data));
  }

  public append(data: T): void {
    this.appendLink(new CircularListNode<T>(data));
  }

  public removeAll(): void {
    let link = this.remove() as CircularListNode<T> | null;
    while (link !== null) {
      link.nextLink = null;
      link = this.remove() as CircularListNode<T> | null;
    }
  }

  public override clear(): void {
    super.clear();
  }

  public baseList(): CircularListBase {
    return this;
  }
}
